runtime/src/installed_scheduler_pool.rs

```diff
@@ -107,15 +145,261 @@ impl BankWithScheduler {
         }
     }
 
+    // take needless &mut only to communicate its semantic mutability to humans...
+    #[cfg(feature = "dev-context-only-utils")]
+    pub fn drop_scheduler(&mut self) {
+        self.inner.drop_scheduler();
+    }
+
+    pub(crate) fn wait_for_paused_scheduler(bank: &Bank, scheduler: &InstalledSchedulerRwLock) {
+        let maybe_result_with_timings = BankWithSchedulerInner::wait_for_scheduler(
+            bank,
+            scheduler,
+            WaitReason::PausedForRecentBlockhash,
+        );
+        assert!(
+            maybe_result_with_timings.is_none(),
+            "Premature result was returned from scheduler after paused"
+        );
+    }
+
+    #[must_use]
+    pub fn wait_for_completed_scheduler(&self) -> Option<ResultWithTimings> {
+        BankWithSchedulerInner::wait_for_scheduler(
+            &self.inner.bank,
+            &self.inner.scheduler,
+            WaitReason::TerminatedToFreeze,
+        )
+    }
+
     pub const fn no_scheduler_available() -> InstalledSchedulerRwLock {
         RwLock::new(None)
     }
 }
 
+impl BankWithSchedulerInner {
+    #[must_use]
+    fn wait_for_completed_scheduler_from_drop(&self) -> Option<ResultWithTimings> {
+        Self::wait_for_scheduler(
+            &self.bank,
+            &self.scheduler,
+            WaitReason::DroppedFromBankForks,
+        )
+    }
+
+    #[must_use]
+    fn wait_for_scheduler(
+        bank: &Bank,
+        scheduler: &InstalledSchedulerRwLock,
+        reason: WaitReason,
+    ) -> Option<ResultWithTimings> {
+        debug!(
+            "wait_for_scheduler(slot: {}, reason: {:?}): started...",
+            bank.slot(),
+            reason,
+        );
+
+        let mut scheduler = scheduler.write().unwrap();
+        let result_with_timings = if scheduler.is_some() {
+            let result_with_timings = scheduler
+                .as_mut()
+                .and_then(|scheduler| scheduler.wait_for_termination(&reason));
+            if !reason.is_paused() {
+                drop(scheduler.take().expect("scheduler after waiting"));
```

Comment:
this meaningless `drop` will be replaced with `SchedulerPool::return_to_pool()` by a later pr.